core/peersholder/peersHolder.go

```diff
@@ -0,0 +1,181 @@
+package peersholder
+
+import (
+	"sync"
+
+	"github.com/ElrondNetwork/elrond-go/core"
+)
+
+type peerInfo struct {
+	pid     core.PeerID
+	shardID uint32
+}
+
+type peersHolder struct {
+	pubKeysToPeerIDs map[string]*peerInfo
+	peersIDsPerShard map[uint32][]core.PeerID
+	peerIDs          map[core.PeerID]struct{}
+	sync.RWMutex
+}
+
+// NewPeersHolder returns a new instance of peersHolder
+func NewPeersHolder(preferredPublicKeys [][]byte) *peersHolder {
+	pubKeysToPeerIDs := make(map[string]*peerInfo)
+
+	for _, pubKey := range preferredPublicKeys {
+		pubKeysToPeerIDs[string(pubKey)] = nil
+	}
+
+	return &peersHolder{
+		pubKeysToPeerIDs: pubKeysToPeerIDs,
+		peersIDsPerShard: make(map[uint32][]core.PeerID),
+		peerIDs:          make(map[core.PeerID]struct{}),
+	}
+}
+
+// Put will perform the insert or the upgrade operation if the provided public key is inside the preferred peers list
+func (ph *peersHolder) Put(pubKey []byte, peerID core.PeerID, shardID uint32) {
+	ph.Lock()
+	defer ph.Unlock()
+
+	pubKeyStr := string(pubKey)
+
+	pInfo, pubKeyExists := ph.pubKeysToPeerIDs[pubKeyStr]
+	if !pubKeyExists {
+		// early exit - the public key is not a preferred peer
+		return
+	}
+
+	if pInfo == nil {
+		ph.addPeerInMaps(pubKeyStr, peerID, shardID)
+		return
+	}
+
+	ph.updatePeerInMaps(pubKeyStr, peerID, shardID)
+}
+
+// Get will return a map containing the preferred peer IDs, split by shard ID
+func (ph *peersHolder) Get() (map[uint32][]core.PeerID, error) {
+	ph.RLock()
+	defer ph.RUnlock()
+
+	if len(ph.peersIDsPerShard) == 0 {
+		return nil, core.ErrEmptyPreferredPeersList
```

Comment:
debatable, I would have returned the empty map.

Reply:
done